File: javascript/packages/linter/src/rules/html-input-require-autocomplete.ts
import { getTagName } from "@herb-tools/core"
import { BaseRuleVisitor, getAttribute, getAttributeValue, getStaticAttributeValueContent } from "./rule-utils.js"
import { ParserRule } from "../types.js"

import type { LintOffense, LintContext } from "../types.js"
import type { ParseResult, HTMLOpenTagNode } from "@herb-tools/core"

class HTMLInputRequireAutocompleteVisitor extends BaseRuleVisitor {
  readonly HTML_INPUT_TYPES_REQUIRING_AUTOCOMPLETE = [
    "color",
    "date",
    "datetime-local",
    "email",
    "month",
    "number",
    "password",
    "range",
    "search",
    "tel",
    "text",
    "time",
    "url",
    "week",
  ]

  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.checkInputTag(node)
  }

  private checkInputTag(node: HTMLOpenTagNode): void {
    if (!this.isInputTag(node) || this.hasAutocomplete(node)) return

    const typeAttribute = getAttribute(node, "type");
    if (!typeAttribute) return

    const typeValue = getStaticAttributeValueContent(typeAttribute)
    if (!typeValue) return

    if (!this.HTML_INPUT_TYPES_REQUIRING_AUTOCOMPLETE.includes(typeValue)) return

    this.addOffense(
      "Input tag is missing an autocomplete attribute. If no autocomplete behaviour is desired, use the value `off` or `nope`.",
      typeAttribute.location
    )
  }

  private hasAutocomplete(node: HTMLOpenTagNode) {
    const autocompleteAttribute = getAttribute(node, "autocomplete");
    if (!autocompleteAttribute) return false

    const autocompleteValue = getAttributeValue(autocompleteAttribute)
    if (!autocompleteValue) return false

    return true
  }

  private isInputTag(node: HTMLOpenTagNode) {
    const tagName = getTagName(node);

    if (tagName === "input") {
      return true
    } else {
      return false
    }
  }
}

export class HTMLInputRequireAutocompleteRule extends ParserRule {
  name = "html-input-require-autocomplete"

  check(result: ParseResult, context?: Partial<LintContext>): LintOffense[] {
    const visitor = new HTMLInputRequireAutocompleteVisitor(this.name, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_input_require_autocomplete.rs
use std::collections::HashSet;
use std::sync::LazyLock;

use crate::utils::tag_utils::{get_attribute, get_static_attribute_value, get_tag_name_from_open_tag};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(InputRequireAutocompleteVisitor);
define_parser_rule!(
  HTMLInputRequireAutocompleteRule,
  "html-input-require-autocomplete",
  Error,
  InputRequireAutocompleteVisitor
);

static INPUT_TYPES_REQUIRING_AUTOCOMPLETE: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
  [
    "color",
    "date",
    "datetime-local",
    "email",
    "month",
    "number",
    "password",
    "range",
    "search",
    "tel",
    "text",
    "time",
    "url",
    "week",
  ]
  .into_iter()
  .collect()
});

impl Visitor for InputRequireAutocompleteVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if get_tag_name_from_open_tag(node).map(|n| n.eq_ignore_ascii_case("input")).unwrap_or(false) {
      let has_autocomplete = get_attribute(node, "autocomplete")
        .and_then(|attribute| get_static_attribute_value(attribute))
        .is_some();

      if !has_autocomplete {
        if let Some(type_value) = get_attribute(node, "type").and_then(|attribute| get_static_attribute_value(attribute)) {
          if INPUT_TYPES_REQUIRING_AUTOCOMPLETE.contains(type_value.as_str()) {
            self.add_offense(
              "Add an `autocomplete` attribute to improve form accessibility. Use a specific value (e.g., `autocomplete=\"email\"`), `autocomplete=\"on\"` for defaults, or `autocomplete=\"off\"` to disable.".to_string(),
              node.location.clone(),
            );
          }
        }
      }
    }
  }
}
